ouroboros-consensus-cardano/tools/db-validator/Main.hs

```diff
@@ -0,0 +1,263 @@
+import           Cardano.Crypto (Hash, ProtocolMagicId (..),
+                     RequiresNetworkMagic (..), decodeAbstractHash)
+
+import qualified Shelley.Spec.Ledger.PParams as SL
+
+import           Ouroboros.Consensus.Block
+import qualified Ouroboros.Consensus.Fragment.InFuture as InFuture
+import qualified Ouroboros.Consensus.Node as Node
+import           Ouroboros.Consensus.Node.DbMarker
+import           Ouroboros.Consensus.Node.ProtocolInfo (ProtocolInfo (..))
+import           Ouroboros.Consensus.Util.IOLike (atomically)
+import           Ouroboros.Consensus.Util.Orphans ()
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import qualified Ouroboros.Consensus.Storage.ChainDB as ChainDB
+import           Ouroboros.Consensus.Storage.ChainDB.Impl.Args (fromChainDbArgs)
+import qualified Ouroboros.Consensus.Storage.ChainDB.Impl.ImmDB as ImmDB
+
+import           Ouroboros.Consensus.Byron.Ledger (ByronBlock)
+import           Ouroboros.Consensus.Byron.Node
+
+import           Ouroboros.Consensus.Shelley.Ledger.Block (ShelleyBlock)
+import           Ouroboros.Consensus.Shelley.Node (Nonce (..), ShelleyGenesis,
+                     protocolInfoShelley)
+import           Ouroboros.Consensus.Shelley.Protocol.Crypto
+
+import           Ouroboros.Consensus.Cardano.Block (CardanoBlock)
+import           Ouroboros.Consensus.Cardano.Node (TriggerHardFork (..),
+                     protocolInfoCardano)
+
+instance ParseField UTCTime
+
+instance ParseFields UTCTime
+
+instance ParseRecord UTCTime where
+  parseRecord = fmap getOnly parseRecord
+
+instance ParseField (Hash CB.Raw) where
+  readField = Options.eitherReader (first Text.unpack . decodeAbstractHash . Text.pack)
+
+instance ParseFields (Hash CB.Raw)
+
+instance ParseRecord (Hash CB.Raw) where
+  parseRecord = fmap getOnly parseRecord
+
+instance ParseField  BlockType
+instance ParseRecord BlockType
+instance ParseFields BlockType
+
+data Args w = Args {
+    dbDir                :: w ::: FilePath        <?> "Path to the new database directory"
+  , configFile           :: w ::: [FilePath]      <?> "Configuration file (e.g. mainnet-genesis.json)"
+  , systemStart          :: w ::: Maybe UTCTime   <?> "System start time"
+  , requiresNetworkMagic :: w ::: Bool            <?> "Expecto patronum?"
+  , genesisHash          :: w ::: Hash CB.Raw     <?> "Expected genesis hash"
+  , blockType            :: w ::: Maybe BlockType <?> "type of network"
+  , verbose              :: w ::: Bool            <?> "Enable verbose logging"
+  , onlyImmDB            :: w ::: Bool            <?> "Validate only the immutable DB (e.g. do not do ledger validation)"
+  } deriving (Generic)
+
+instance ParseRecord (Args Wrapped)
+
+deriving instance Show (Args Unwrapped)
+
+data BlockType = Byron | Shelley | Cardano
+  deriving (Generic, Show, Read)
+
+main :: IO ()
+main = do
+  (cmd :: Args Unwrapped) <- unwrapRecord "DB validator"
+  let Args{ dbDir, blockType } = cmd
```

Comment:
Same as above, can these lines be merged without exceeding 80 chars?

Reply:
I couldn't actually find a syntax to make this one line.

Reply:
What about:
```haskell
cmd@Args { dbDir, blockType } :: Args Unwrapped <- unwrapRecord "DB validator"
```

Ah, that will be too long, so never mind.